### PR TITLE
Fix data type comparisons

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
Fixes dataype comparions for normal Pointer instances.
(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [ X] Relevant tests for your changes are passing.

### CODE REVIEW

File: libnd4j/include/legacy/cuda/NativeOps.cu [hunk 1]
```diff
@@ -3565,11 +3565,19 @@ void setShapeBuffer(sd::LongType *inputShapeData,sd::DataType dt,sd::LongType *b
...
+    descriptor._extraProperties = ARRAY_EMPTY;
+    sd_printf("Setting empty for shape buffer \n",0);
+  }
+
+  auto buffer = descriptor.toShapeInfo();
   for(sd::LongType i = 0; i < len; i++) {
     bufferToSet[i] = buffer[i];
   }
 
+  sd_printf("Shape buffer is empty: %d\n",shape::isEmpty(buffer));
```
Comment (treo) [inline libnd4j/include/legacy/cuda/NativeOps.cu:3579]: leftover print.

File: libnd4j/include/ops/declarable/impl/DeclarableOp.cpp [hunk 2]
```diff
@@ -540,6 +551,7 @@ sd::Status sd::ops::DeclarableOp::validateDataTypes(Context &block) {
       index++;
     }
   } else {
+    sd_printf("Not Fast path out\n",0);
```
Comment (treo) [inline libnd4j/include/ops/declarable/impl/DeclarableOp.cpp]: leftover print

File: libnd4j/include/array/ShapeDescriptor.h [hunk 3]
```diff
@@ -74,6 +66,14 @@ class SD_LIB_EXPORT ShapeDescriptor {
...
 
  public:
+  int _rank = 0;
+  std::vector<sd::LongType> _shape_strides;
+  // std::vector<sd::LongType> _strides;
+  sd::LongType _ews = 1;
+  char _order = 'c';
+  DataType _dataType;
+  sd::LongType _extraProperties = 0;
+  sd::LongType _paddedAllocSize = 0;
```
Comment (treo) [inline libnd4j/include/array/ShapeDescriptor.h:76]: Does this have to be public? Or is that left over from debugging?

**Review by treo — CHANGES_REQUESTED** [review]

File: libnd4j/include/array/ShapeDescriptor.h (hunk 3 above)
Comment (agibsonccc) [reply libnd4j/include/array/ShapeDescriptor.h:76]: Yes. This was orignally private but I ended up needing to use extraProperties in a  few places.

File: libnd4j/include/legacy/impl/cnpy.cpp [hunk 4]
```diff
@@ -384,10 +384,20 @@ cnpy::NpyArray cnpy::loadNpyFromHeader(char *data) {
...
+      std::string header;
+      for(int i = 0; i < hdr.size(); i++) {
+        header+= hdr[i];
+      }
 
+      firstError += header;
+      throw std::runtime_error(firstError);
+    }
+  } else {
+    throw std::runtime_error("ointer doesn't look like a NumPy header. Missing expected character at first value.");
```
Comment (treo) [inline libnd4j/include/legacy/impl/cnpy.cpp]: Typo.

**Review by treo — CHANGES_REQUESTED** [review]
One more typo to fix, otherwise it looks ok.

File: libnd4j/include/legacy/impl/cnpy.cpp (hunk 4 above)
Comment (agibsonccc) [reply libnd4j/include/legacy/impl/cnpy.cpp]: Fixed

**Review by treo — APPROVED** [review]

**Pull request merged**